bindings/java/devdoc/java_out_of_process_hld.md

```diff
@@ -0,0 +1,105 @@
+Java Out of Process Module API
+==============================
+
+Overview
+--------
+
+This document describes the API that enables communication between an out of process Java module and Azure IoT Gateway. 
+An out of process module should implement `IGatewayModule` interface or extend the `GatewayModule` abstract class as it does for an in-process module, more details in [Java binding high level design](java_binding_hld.md). In addition, the API provides `GatewayModuleProxy` which is a proxy for the IGatewayModule implementation and handles the communication between the Gateway and the out of process module. 
+`GatewayModuleProxy` contains the methods to *attach* a module to the Gateway which will start listening for messages and *detach* from the Gateway. 
+
+There are two types of messages that are exchanged between the Gateway and the out of process module:
+
+-  *control* messages that are used to signal `CREATE`, `START` and `DESTROY` of a module. These messages are always sent from the Gateway to the module.
+-  *data* messages which are sent between modules via the Gateway.
+
+
+![](java_oop_module_hld.png)
+
+Receive out of process messages from the Gateway
+------------------------------------------------
+
+In order to make a module to run out of process, create an instance of `GatewayModuleProxy`, pass the configuration that contains the unique identifier to attach to the Gateway and then call `attach`. 
+Also the Gateway configuration has to be changed for that specific module to set the loader as an out of process loader and set the same unique identifier. 
+
+``` java
+interface ModuleProxy {
```

Comment:
Dave has already used this name (proxy module) for the gateway-side module. I was using OopModule for out-of-process module.

I chose OopModule, because a person wanting to use the Gateway as it is today would `#include module.h` in their library. Now, if they include `oop_module.h` instead (effectively prefixing `module.h` with `oop_`) and creating a `main` that will call `attach` and `detach`, then the library becomes an out-of-process module.

I'm not sure how it works in Java, but perhaps you can do something similar with your naming.